pkg/destroy/powervs/powervs.go

```diff
@@ -739,6 +732,83 @@ func (o *ClusterUninstaller) loadSDKServices() error {
 	return nil
 }
 
+// ServiceInstanceNameToGUID returns the GUID of the matching service instance name which was passed in.
+func (o *ClusterUninstaller) ServiceInstanceNameToGUID(ctx context.Context, name string) (string, error) {
+	var (
+		options   *resourcecontrollerv2.ListResourceInstancesOptions
+		resources *resourcecontrollerv2.ResourceInstancesList
+		err       error
+		perPage   int64 = 10
+		moreData        = true
+		nextURL   *string
+		groupID   = o.resourceGroupID
+	)
+
+	// If the user passes in a human readable group id, then we need to convert it to a UUID
+	listGroupOptions := o.managementSvc.NewListResourceGroupsOptions()
+	groups, _, err := o.managementSvc.ListResourceGroupsWithContext(ctx, listGroupOptions)
+	if err != nil {
+		return "", fmt.Errorf("failed to list resource groups: %w", err)
+	}
+	for _, group := range groups.Resources {
+		if *group.Name == groupID {
+			groupID = *group.ID
+		}
+	}
+
+	options = o.controllerSvc.NewListResourceInstancesOptions()
+	options.SetResourceGroupID(groupID)
+	// resource ID for Power Systems Virtual Server in the Global catalog
+	options.SetResourceID(powerIAASResourceID)
+	options.SetLimit(perPage)
+
+	for moreData {
+		resources, _, err = o.controllerSvc.ListResourceInstancesWithContext(ctx, options)
+		if err != nil {
+			return "", fmt.Errorf("failed to list resource instances: %w", err)
+		}
+
+		for _, resource := range resources.Resources {
+			var (
+				getResourceOptions *resourcecontrollerv2.GetResourceInstanceOptions
+				resourceInstance   *resourcecontrollerv2.ResourceInstance
+				response           *core.DetailedResponse
+			)
+
+			getResourceOptions = o.controllerSvc.NewGetResourceInstanceOptions(*resource.ID)
+
+			resourceInstance, response, err = o.controllerSvc.GetResourceInstance(getResourceOptions)
+			if err != nil {
+				return "", fmt.Errorf("failed to get instance: %w", err)
+			}
+			if response != nil && response.StatusCode == gohttp.StatusNotFound || response.StatusCode == gohttp.StatusInternalServerError {
+				return "", fmt.Errorf("failed to get instance, response is: %v", response)
```

Comment:
There is a chance that you'll fail to get some resource instances but you can still find the one you're looking for. So you can increase the chances of success by saving the last error you saw but keep going through the resources. Then, at the end of the loop, you can return the last error found. Of course it's OK If you prefer to fail right away.

Reply:
Hrm.  Destroy code is written to fail at the first sign of trouble.  I do agree in this scenario, that ignoring an unrelated and very transitory hiccup while searching for the correct service instance could make this code more resilient.  In the long run, with PowerVS, a problem found usually indicates more problems down the road.

Reply:
@mjturek We should have a TODO to mention in the documentation that destroy code will fail at the first sign of trouble.  And that the customer should open a support ticket and try again after support responds.